go/parquet/metadata/metadata_test.go

```diff
@@ -0,0 +1,325 @@
+	rg2Builder := fbuilder.AppendRowGroup()
+	col1Builder = rg2Builder.NextColumnChunk()
+	col2Builder = rg2Builder.NextColumnChunk()
+	// column metadata
+	col1Builder.SetStats(statsInt)
+	col2Builder.SetStats(statsFloat)
+	col1Builder.Finish(nrows/2, 6, 0, 10, 512, 600, true, false, dictEncodingStats, dataEncodingStats, nil)
+	col2Builder.Finish(nrows/2, 16, 0, 26, 512, 600, true, false, dictEncodingStats, dataEncodingStats, nil)
+
+	rg2Builder.SetNumRows(nrows / 2)
+	rg2Builder.Finish(1024, -1)
+
+	return fbuilder.Finish()
+}
+
+func assertStatsSet(t *testing.T, m *metadata.ColumnChunkMetaData) {
+	ok, err := m.StatsSet()
+	assert.NoError(t, err)
+	assert.True(t, ok)
+}
+
+func assertStats(t *testing.T, m *metadata.ColumnChunkMetaData) metadata.TypedStatistics {
+	s, err := m.Statistics()
+	assert.NoError(t, err)
+	assert.NotNil(t, s)
+	return s
+}
+
+func TestBuildAccess(t *testing.T) {
+	props := parquet.NewWriterProperties(parquet.WithVersion(parquet.V2))
+
+	fields := schema.FieldList{
+		schema.NewInt32Node("int_col", parquet.Repetitions.Required, -1),
+		schema.NewFloat32Node("float_col", parquet.Repetitions.Required, -1),
+	}
+	root, err := schema.NewGroupNode("schema", parquet.Repetitions.Repeated, fields, -1)
+	require.NoError(t, err)
+	schema := schema.NewSchema(root)
+
+	var (
+		nrows      int64   = 1000
+		intMin     int32   = 100
+		intMax     int32   = 200
+		floatMin   float32 = 100.100
+		floatMax   float32 = 200.200
+		statsInt   metadata.EncodedStatistics
+		statsFloat metadata.EncodedStatistics
+	)
+
+	statsInt.SetNullCount(0).
+		SetDistinctCount(nrows).
+		SetMin((*(*[4]byte)(unsafe.Pointer(&intMin)))[:]).
+		SetMax((*(*[4]byte)(unsafe.Pointer(&intMax)))[:])
+
+	statsFloat.SetNullCount(0).
+		SetDistinctCount(nrows).
+		SetMin((*(*[4]byte)(unsafe.Pointer(&floatMin)))[:]).
+		SetMax((*(*[4]byte)(unsafe.Pointer(&floatMax)))[:])
+
+	faccessor, err := generateTableMetaData(schema, props, nrows, statsInt, statsFloat)
+	require.NoError(t, err)
+	serialized, err := faccessor.SerializeString(context.Background())
+	assert.NoError(t, err)
+	faccessorCopy, err := metadata.NewFileMetaData([]byte(serialized), nil)
+	assert.NoError(t, err)
+
+	for _, accessor := range []*metadata.FileMetaData{faccessor, faccessorCopy} {
+		assert.Equal(t, nrows, accessor.NumRows)
+		assert.Len(t, accessor.RowGroups, 2)
+		assert.EqualValues(t, parquet.V2, accessor.Version)
+		assert.Equal(t, parquet.DefaultCreatedBy, accessor.GetCreatedBy())
+		assert.Equal(t, 3, accessor.NumSchemaElements())
+
+		rg1Access := accessor.RowGroup(0)
+		assert.Equal(t, 2, rg1Access.NumColumns())
+		assert.Equal(t, nrows/2, rg1Access.NumRows())
+		assert.Equal(t, int64(1024), rg1Access.TotalByteSize())
+
+		rg1Col1, err := rg1Access.ColumnChunk(0)
+		assert.NoError(t, err)
+		rg1Col2, err := rg1Access.ColumnChunk(1)
+		assert.NoError(t, err)
+		assertStatsSet(t, rg1Col1)
+		assertStatsSet(t, rg1Col2)
+		assert.Equal(t, statsInt.Min, assertStats(t, rg1Col1).EncodeMin())
```

Comment:
its a little hard to grok/review if all of these tests are correct.  I'm glad there are tests, but I'm wondering if there might be a better way to express the tests so that is it easier to verify the assertions align with reality?

Reply:
I actually copied these test cases from the C++ implementation (though that was back in january). I'll see if i can make the tests more obvious as to what they are testing.